lib/routes-priv.js

```diff
@@ -60,6 +67,105 @@ class RoutesPriv {
     this.websocketOpts = { requestTimeout: requestTimeout * 1000 }
   }
 
+  onPrivateTosSetHttpRequest (req, res) {
+    this.validateUser(req, res, (err, data) => {
+      if (err) {
+        return this.handleAuthError(err, res)
+      }
+
+      const { actor } = data.parseReq
+
+      const meta = data.verify.meta
+      const user = meta.actions[0].authorization[0].actor
+      if (actor !== user) {
+        log('ERR_ACTOR_USER', actor, user)
+        return res.status(403).json({ error: 'ERR_INVALID_PAYLOAD' })
+      }
+
+      const ts = Date.now()
+      this.tosCollection.insertOne({
+        _id: TOS + '_' + user,
+        ts: ts,
+        user: user
+      }, (err) => {
+        if (err && err.code === 11000) {
+          return res.status(200).json({ ok: true })
+        }
+
+        if (err) {
+          log(err)
+          return sendGenericError(res)
+        }
+
+        return res.status(200).json({ ok: true })
+      })
+    })
+  }
+
+  onPrivateTosGetHttpRequest (req, res) {
+    this.validateUser(req, res, (err, data) => {
+      if (err) {
+        return this.handleAuthError(err, res)
+      }
+
+      const { actor } = data.parseReq
+
+      const meta = data.verify.meta
+      const user = meta.actions[0].authorization[0].actor
+      if (actor !== user) {
+        log('ERR_ACTOR_USER', actor, user)
+        return res.status(403).json({ error: 'ERR_INVALID_PAYLOAD' })
+      }
+
+      const key = TOS + '_' + user
+      this.tosCollection.findOne({ _id: key }, (err, data) => {
+        if (err) {
+          log(err)
+          return sendGenericError(res)
+        }
+
+        const signed = data && data._id === key
+        return res.status(200).json({ signed_tos: signed || false })
+      })
+    })
+  }
+
+  onPrivateHistoryHttpRequest (req, res) {
+    const { maxResults } = this.conf
+
+    this.validateUser(req, res, (err, data) => {
+      if (err) {
+        return this.handleAuthError(err, res)
+      }
+
+      const { actor, payload } = data.parseReq
+
+      const meta = data.verify.meta
+      const limit = Math.min(payload.limit, maxResults)
+      const user = meta.actions[0].authorization[0].actor
+      const stmt = { username: user }
+
+      if (actor !== user) {
+        log('ERR_ACTOR_USER', actor, user)
+        return res.status(403).json({ error: 'ERR_INVALID_PAYLOAD' })
+      }
+
+      this.tradesCollection
+        .find(stmt, { limit: limit })
+        .sort({ ts: -1 })
```

Comment:
if going to sort by ts is important to have:
db.compliances.createIndex({ ts: -1 })

Reply:
Indexes stated here, they are fine
https://github.com/bitfinexcom/moonbeam-history#indexes

Reply:
true, indexes are created via mongo shell.